src/xpcc/driver/storage/block_device_spiflash.hpp

```diff
@@ -0,0 +1,230 @@
+// coding: utf-8
+/* Copyright (c) 2017, Raphael Lehmann
+ * All Rights Reserved.
+ *
+ * The file is part of the xpcc library and is released under the 3-clause BSD
+ * license. See the file `LICENSE` for the full license governing this code.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC_BLOCK_DEVICE_SPIFLASH_HPP
+#define XPCC_BLOCK_DEVICE_SPIFLASH_HPP
+
+#include <xpcc/architecture/interface/block_device.hpp>
+
+#include <xpcc/architecture/interface/register.hpp>
+#include <xpcc/architecture/interface/spi_device.hpp>
+#include <xpcc/processing/resumable.hpp>
+
+namespace xpcc
+{
+
+
+/**
+ * \brief	Block device with SPI Flash (Microchip SST26VF064B)
+ *
+ * 64MBit flash chip in SOIJ-8, WDFN-8 or SOIC-16
+ *
+ * \tparam Spi			The SpiMaster interface
+ * \tparam Cs			The GpioOutput pin connected to the flash chip select
+ * \tparam flashSize	Flash chip size in byte
+ *
+ * The `read()`, `erase()`,`program()` and `write()` methodes wait for
+ * the chip to finish writing to the flash.
+ *
+ * \ingroup	driver_storage
+ * \author	Raphael Lehmann
+ */
+template <typename Spi, typename Cs, size_t flashSize>
```

Comment:
On AVR size_t is a 16 bit unsigned int. This wouldn't allow you to use a flash larger than 65535 bytes. You could take a plain uint32_t or something like `using flash_size_t = ...` and define it to uint32_t on microcontrollers and size_t on hosted to also support 4GB+ files. The same applies for the size_t function parameters.

Reply:
> This wouldn't allow you to use a flash larger than 65535 bytes.

No. You are just unable to write, read or delete >64k memory at once. `address` is a `uint32_t`, so you can write/read to/from a <=4GBit flash chip in <=64kBit blocks.

For write and read operations you would need a >64k buffer in your AVRs RAM, but erasing a complete 64MBit flash chip at once on AVR is more realistic. Hmmm...

Reply:
Sorry, I missed that this `size_t` is the template parameter and not an read/write/erase function argument.

I'll change the `flashSize` type to `uint32_t` here.